Realm/RLMQueryUtil.mm

```diff
@@ -980,6 +989,34 @@ void update_query_with_column_expression(RLMSchema *schema, RLMObjectSchema *des
                             std::move(left), std::move(right));
 }
 
+NSExpression *simplify_self_value_for_key_path_function_expression(NSExpression *expression) {
+    // -[NSPredicate predicateWithSubtitutionVariables:] results in function expressions of the form [SELF valueForKeyPath:]
+    // that update_query_with_predicate cannot handle. Replace such expressions with equivalent NSKeyPathExpressionType expressions.
+    if (expression.expressionType == NSFunctionExpressionType
+        && expression.operand.expressionType == NSEvaluatedObjectExpressionType
+        && [expression.function isEqualToString:@"valueForKeyPath:"]) {
+        if (NSString *keyPath = expression.arguments.firstObject.keyPath) {
```

Comment:
I addressed both this and the other gnarly if statement in b227714 by introducing helper functions. It's more code, but a lot easier to follow.